http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala

```diff
@@ -0,0 +1,367 @@
+      )
+    } yield ()
+
+  def pollCheck(googleProject: GoogleProject,
+                runtimeAndRuntimeConfig: RuntimeAndRuntimeConfig,
+                operation: com.google.cloud.compute.v1.Operation,
+                action: RuntimeStatus)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit]
+
+  private def handler(monitorContext: MonitorContext, monitorState: MonitorState): F[CheckResult] =
+    for {
+      now <- nowInstant
+      implicit0(ct: ApplicativeAsk[F, AppContext]) = ApplicativeAsk.const[F, AppContext](
+        AppContext(monitorContext.traceId, now)
+      )
+      r <- monitorState match {
+        case MonitorState.Initial => handleInitial(monitorContext)
+        case MonitorState.CheckTools(ip, runtimeAndRuntimeConfig, images, _) =>
+          handleCheckTools(monitorContext, runtimeAndRuntimeConfig, ip, Set.empty, images)
+        case MonitorState.Check(runtimeAndRuntimeConfig) => handleCheck(monitorContext, runtimeAndRuntimeConfig)
+      }
+    } yield r
+
+  def failedRuntime(monitorContext: MonitorContext,
+                    runtimeAndRuntimeConfig: RuntimeAndRuntimeConfig,
+                    errorDetails: Option[RuntimeErrorDetails],
+                    instances: Set[DataprocInstance])(
+    implicit ev: ApplicativeAsk[F, AppContext]
+  ): F[CheckResult]
+
+  def readyRuntime(runtimeAndRuntimeConfig: RuntimeAndRuntimeConfig,
+                   publicIp: IP,
+                   monitorContext: MonitorContext,
+                   dataprocInstances: Set[DataprocInstance]): F[CheckResult] =
+    for {
+      now <- nowInstant
+      _ <- persistInstances(runtimeAndRuntimeConfig, dataprocInstances)
+      // update DB after auth futures finish
+      _ <- clusterQuery.setToRunning(runtimeAndRuntimeConfig.runtime.id, publicIp, now).transaction
+      // Record metrics in NewRelic
+      _ <- RuntimeMonitor.recordStatusTransitionMetrics(
+        monitorContext.start,
+        RuntimeMonitor.getRuntimeUI(runtimeAndRuntimeConfig.runtime),
+        runtimeAndRuntimeConfig.runtime.status,
+        RuntimeStatus.Running,
+        runtimeAndRuntimeConfig.runtimeConfig.cloudService
+      )
+      _ <- if (runtimeAndRuntimeConfig.runtime.status == RuntimeStatus.Creating)
+        RuntimeMonitor.recordClusterCreationMetrics(
+          runtimeAndRuntimeConfig.runtime.auditInfo.createdDate,
+          runtimeAndRuntimeConfig.runtime.runtimeImages,
+          monitorConfig.imageConfig,
+          runtimeAndRuntimeConfig.runtimeConfig.cloudService
+        )
+      else F.unit
+      timeElapsed = (now.toEpochMilli - monitorContext.start.toEpochMilli).milliseconds
+      // Finally pipe a shutdown message to this actor
+      _ <- logger.info(
+        s"${monitorContext} | Runtime ${runtimeAndRuntimeConfig.runtime.projectNameString} is ready for use after ${timeElapsed.toSeconds} seconds!"
+      )
+    } yield ((), None)
+
+  def checkAgain(
```

Comment:
I'm a bit confused as to the purpose of the `checkAgain` method -- can you elaborate?

Reply:
it's used anywhere the transition is not complete yet...before setting the state too either `Check` or `CheckTools` which will cause the stream the sleep some interval and continue, it also checks if we should timeout the monitoring process